src/v/cloud_topics/objects/level_one_object.cc

```diff
@@ -0,0 +1,447 @@
+
+    ss::future<object_info> finish() final {
+        end_partition();
+        auto footer_start = _offset;
+        co_await serde_write_to_stream(data_type::footer, _index.copy());
+        object_info info{
+          .index = std::move(_index),
+          .footer_offset = footer_start,
+          .size_bytes = _offset + sizeof(uint32_t),
+        };
+        auto footer_size
+          = std::bit_cast<std::array<char, sizeof(uint32_t)>, uint32_t>(
+            _offset - footer_start);
+        co_await _output.write(footer_size.data(), footer_size.size());
+        co_return info;
+    }
+
+    ss::future<> close() final { return _output.close(); }
+
+private:
+    void end_partition() {
+        if (_current_partition.ntp == model::ntp{}) {
+            return;
+        }
+        _index.partitions.push_back(std::exchange(_current_partition, {}));
+    }
+
+    template<typename T>
+    ss::future<> serde_write_to_stream(data_type dt, T data) {
+        iobuf b;
+        b.append(as_bytes(dt));
+        iobuf serialized;
+        co_await serde::write_async(serialized, std::move(data));
+        b.append(as_bytes<uint32_t>(serialized.size_bytes()));
+        b.append(std::move(serialized));
+        _offset += b.size_bytes();
+        co_await write_iobuf_to_output_stream(std::move(b), _output);
+    }
+
+    ss::future<> write_batch_to_stream(model::record_batch batch) {
+        iobuf b;
+        b.append(as_bytes(data_type::kafka_batch));
+        for_each_batch_header_field(
+          batch.header(), [&b](auto arg) { b.append(as_bytes(arg)); });
+        b.append(std::move(batch).release_data());
+        _offset += b.size_bytes();
+        return write_iobuf_to_output_stream(std::move(b), _output);
+    }
+
+    size_t _offset = 0;
+    ss::output_stream<char> _output;
+    object_index _index;
+    object_index::partition _current_partition;
+    options _opts;
+};
+
+} // namespace
+
+std::unique_ptr<object_builder>
+object_builder::create(ss::output_stream<char> output, options opts) {
+    return std::make_unique<object_builder_impl>(std::move(output), opts);
+}
+
+namespace {
+
+class object_reader_impl : public object_reader {
+public:
+    explicit object_reader_impl(ss::input_stream<char> input)
+      : _input(std::move(input)) {}
+
+    ss::future<result> read_next() final {
+        auto dt_buf = co_await _input.read_exactly(sizeof(data_type));
+        if (dt_buf.size() != sizeof(data_type)) {
+            throw std::runtime_error(fmt::format(
+              "expected {} bytes for data type, got: {}",
+              sizeof(data_type),
+              dt_buf.size()));
+        }
+        switch (from_bytes<data_type>(dt_buf.get())) {
+        case data_type::kafka_batch:
+            co_return co_await read_next_batch();
+        case data_type::partition_marker:
+            co_return co_await read_next_serde<model::ntp>();
+        case data_type::footer:
+            co_return co_await read_next_serde<object_index>();
```

Comment:
Given this will be based on stored data, add a default case that throws?

Reply:
Good call, done here: https://github.com/redpanda-data/redpanda/commit/77e7605c0b37ebdcb3cc18e3ad85a57f83cd0bd9